documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+    public Joysticks Joysticks { get; }
+    public IReadOnlyList<IInputDevice> Devices { get; }
+    public List<IInputBackend> Backends { get; }
+    public void Update();
+}
+```
+
+The central input object acts as the main entry point into the Input API, and is responsible for comparing the state reported by the devices for differences between `Update` calls (raising events as necessary). 
+
+`Mice`, `Keyboards`, `Gamepads`, and `Joysticks` are all custom `IReadOnlyList` types for enumerating the **state** (given that the actual input objects are thin, and the state contains a reference to the original device anyway). However, these custom types also contain the events. This is so we can "scope" the events, rather than putting them at the top-level and having to call the events `MouseButtonDown`, `JoystickButtonDown`, etc.
+
+By virtue of the `State` properties not updating until `IInputBackend.Update` is called, the states enumerated by the lists will not change until `Update` is called.
+
+`Update` will call `IInputBackend.Update` on each of the `Backends`, and retrieve the state for each device contained in each backend. As it is doing this, it updates the lists within the input context.
+
+After the lists have been updated, the `Update` method compares the old state and the new state for that device and raises events accordingly. If there is no old state available for a device, the `ConnectionChanged` event (read on!) within each of the custom lists is called with a value of `true` signifying that a device has connected. Likewise, if there is no new state available for a device we have old state for, the `ConnectionChanged` event will be called with a value of `false` signifying that a device has been disconnected.
+
+`Backends` is a mutable list of input backends. Changes to this list do not take effect until `Update` is called again. The `ConnectionChanged` rules above will still be respected e.g. when you remove a backend, all of its devices will have a disconnected event raised for them.
+
+`Devices` contains all devices reported by all backends, including devices that do not necessarily fit into one of our more specialized wrapper lists. This means that if a backend has a device type we do not recognise, it will be accessible via this list.
+
+## Custom List Types
+
+These are relatively simple, though note that we explicitly implement the `IReadOnlyList` indexer in favour of one that returns a `ref readonly` type and a corresponding custom enumerator.
+
+The actual type of the custom enumerator at this time the Silk.NET team would like to leave implementation defined, but with the following requirements:
+- The `Next` method must return a `ref readonly` type.
+- The type can be used as an operand in a `foreach` loop
+- While it may not necessarily implement `IEnumerator<T>`, it must implement the surface API as if it does where `T` is a `ref readonly` type (which is forbidden in the language today).  
```

Comment:
Can't we just simplify this to be a ref readonly returning enumerator + "refer to the language rules"

Reply:
```suggestion
The actual type of the custom enumerator at this time the Silk.NET team would like to leave implementation defined, but with the following requirements:
- The `Next` method must return a `ref readonly` type.
- The type can be used as an operand in a `foreach` loop
- This type is an enumerator type that returns a `ref readonly T` as defined by the C# language rules.
```